### ISSUE TITLE
Improve build UX for official release tarball - don't ask users to supply url/sha

### ISSUE THREAD

**Issue by omajid** [dotnet/source-build#3366]
When I download a release tarball for .NET 8 from the VMR (dotnet/dotnet) repository, I would like to be able to build it without providing additional information. 

This came up in https://github.com/dotnet/source-build/issues/3265:

> Modified build.sh to take in a url and sha for the source to use by sourcelink. If not provided require the build is executing in a git context and read the values from it.

An official release tarball doesn't have a git context. It is produced from the ofificial VMR repo and has the same SHA as a the official tag/release. For this tarball, there is only one correct set of values for URL/SHA. Any other value (URL or SHA) will (silently?) produce wrong results for anyone building .NET.

It would be great if everyone consuming the tarball didn't have to figure out what the URL and SHA need to be.

**Comment by MichaelSimons** [comment]
@mmitche, @mthalman, @premun, @tkapin - @omajid  provided this reference link as a reference for how other projects handle specifying revision info - https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_using_forges_hosted_revision_control

**Comment by omajid** [comment]
In case it wasn't clear, https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_using_forges_hosted_revision_control (in particular, the [Multiple Sources Example](https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_multiple_sources_example)) is about how Fedora suggests to the Fedora packagers how to package projects where projects are pulled down from GitHub (or another forge, like GitLab). Both commits and tags can be used to specify what to pull down. Multiple source artifacts (eg, source tarball and auxillary metadata file) are supported, but it's a little bit more painful to consume.

**Comment by premun** [comment]
### Proposal

- When building from a context of a git repo, its remote + current revision is used (so the same as it is now). No override is possible unfortunately (we cannot override this for sourcelink yet).
- When building outside of a context of a git repo, information must be provided to `build.sh`.
  - Remote URI (via `--source-repository`) and SHA of a commit in that remote (`--source-version`)
  - Alternatively, a git tag could be provided via `--source-tag` (e.g. `v8.0.100-preview.3.23178.7`)
  - Alternatively, a metadata file could be provided via `--source-metadata`), see below

#### Release metadata file

Every release on the https://github.com/dotnet/dotnet repo will then (apart from the default tarball) contain a new metadata file named `release.json`. The contents of the file might grow over time with more useful information but the bare minimum it would contain now would be these fields:

```json
{
    "sourceRepository": "https://github.com/dotnet/dotnet",
    "sourceVersion": "6d6f38c2d65c50dfc31dcdadfef1ba24ed4efc45"
}
```

Other possible fields might contain the release name, channel, version, prerelease flag etc.

#### Release description

The release on the https://github.com/dotnet/dotnet repo itself could have a body populated with instructions on building the repo. I imagine something simple like this:

```sh
curl -Lo dotnet-sdk-source-v8.0.100-preview.3.23178.7.tar.gz https://github.com/dotnet/dotnet/archive/refs/tags/v8.0.100-preview.3.23178.7.tar.gz
tar zxf dotnet-sdk-source-v8.0.100-preview.3.23178.7.tar.gz && rm -f dotnet-sdk-source-v8.0.100-preview.3.23178.7.tar.gz
(cd dotnet-8.0.100-preview.3.23178.7 && ./build.sh --source-tag v8.0.100-preview.3.23178.7)
```

We could also have another set of instructions for downloading both the tarball and the metadata file separately and running `build.sh --source-metadata release.json`.

#### Alternatives for `build.sh`

When the metadata file is provided to the `build.sh` script, it will use the remote & SHA from there.

When a git tag is provided, it will pull the metadata file from a release associated with the tag (e.g. `https://github.com/dotnet/dotnet/releases/tag/v8.0.100-preview.3.23178.7/release.json`).

### Questions

1. Format of the file - MSBuild `.props` or `.json`? Or both?
    I guess when we arrive at a point when can pass these properties to sourcelink (if ever), MSBuild might be okay. On the other hand, we're putting a JSON in the release announcement already so that's something our partners use?
2. How would this file be parsed? I understand we can only expect bash (no MSBuild, no `jq`..). I see cases where we grep `.props` files already for instance.
3. Would we document this file somewhere? In the main README? Or is there a suitable source-build `.md`?

cc @MichaelSimons @mthalman @omajid @tkapin @mmitche

**Comment by omajid** [comment]
Sorry about the dely in replying:

> @omajid would it be useful for usability if the GitHub release itself would have a markdown body with copyable instructions including the SHA and the URI?

I am not currently using the data in the release. My current approach (which, to be fair, is changing as the release tarball feature changes) is [just looking for the for a specific release at a sepcific computed location](https://pagure.io/dotnet-sig/dotnet8.0/blob/main/f/dotnet8.0.spec#_77).

> E.g. you either download both the tar.gz or the release metadata file (this would work programmatically), or you could navigate to the release page and copy out the full build command which would have the URI and the SHA.

If we need the release metadata, having it at a well known location for every release (eg, something like `https://github.com/dotnet/dotnet/releases/download/${sdk_or_runtime_version}/metadata`) would be easiest for me.

> From the discussion, I didn't quite understand whether the flow you're doing and that requires you to input these values is a manual dev one or an official pull of the release for package repository purposes.

For my .NET packages for Fedora, I use a script called `update-release`. I call it like this `./update-release $SDK_VERSION $RUNTIME_VERSION`. That [eventually](https://pagure.io/dotnet-sig/dotnet8.0/blob/main/f/build-dotnet-tarball#_122) downloads (and preps, if needed) the source tarball. I would like to extend this to include the metadata file, so if the file is at a well-known path, my script should be able to handle too.

(The Fedora update-to-new-.NET-release is a manual process that a dev runs, and then pushes the result to Fedora's package-build-scripts repository)

> Secondly, were you using the pack-sources.sh script before for your manual dev workflow and we've removed it from you?

I wasn't using it really. I used it for a single preview release for .NET 8 (when it was the recommended way to generate the tarball) and have now stopped using it completely.

Your proposal sounds good to me.

One caveat though:

> When a git tag is provided, it will pull the metadata file from a release associated with the tag (e.g. https://github.com/dotnet/dotnet/releases/tag/v8.0.100-preview.3.23178.7/release.json).

Our builds are offline, so `build.sh` won't be able to talk to github.com. So I suppose that means we have to use the metadata file directly, which should be okay.

**Comment by omajid** [comment]
> How would this file be parsed? I understand we can only expect bash (no MSBuild, no jq..). I see cases where we grep .props files already for instance.

When we are building the SDK, we know we have another .NET SDK available. Could we write a .NET script to parse the file and run it using that already-existing SDK?

**Comment by premun** [comment]
> One caveat though
> Our builds are offline, so build.sh won't be able to talk to github.com. So I suppose that means we have to use the metadata file directly, which should be okay.

Yeah, I see this is a suitable alternative for someone with different restrictions. But I am also open to not having it..

**Comment by tmds** [comment]
> We'd need to get the SHA in there somehow (e.g. in some file) and read that, but we don't know the SHA until after we make the actual commit.

Sourcelink needs a SHA to get the source files. This SHA doesn't need to be the SHA of the tag exactly. So when you tag a release, you can first make a commit that adds the SHA for sourcelink in a file, and then tag that commit. That should work.

**Comment by premun** [comment]
> Sourcelink needs a SHA to get the source files. This SHA doesn't need to be the SHA of the tag exactly. So when you tag a release, you can first make a commit that adds the SHA for sourcelink in a file, and then tag that commit. That should work.

Yeah, that occurred to me but I don't think we want that really. The whole point of the VMR is to have a single commit that identifies everything that's needed for the whole SDK. Doesn't seem right to get a second commit involved? Additionally, it's quite nice for us to have commits of the VMR associated 1:1 with installer it synced from.

I think we also want to make sure people can set whatever information - e.g. they make their own custom build of .NET somewhere in private with some very minimal internal changes but then you'd want to be able to point this to the public dotnet/dotnet repo for anyone using this .NET in public.

**Comment by tmds** [comment]
I think it would be nice if the tarball would _just work_, which is why I wanted to mention it as an option.

> Additionally, it's quite nice for us to have commits of the VMR associated 1:1 with installer it synced from.

I don't know what this means.

> I think we also want to make sure people can set whatever information

We're not precluding that.

**Comment by premun** [comment]
> I think it would be nice if the tarball would just work, which is why I wanted to mention it as an option.

Thanks, it's appreciated!

Maybe my argument about a second commit is not applicable because SourceLink technically always expects this other remote commit somewhere anyway. I am just not sure whether we'd want to always add an extra commit to each we make in the VMR.

**Comment by tmds** [comment]
> to always add an extra commit to each we make in the VMR.

to each release tag.

**Comment by MichaelSimons** [comment]
> > to always add an extra commit to each we make in the VMR.
> 
> for each release tag.

It is not ideal IMO that the sourcelink info would be off on all commits besides the release commit.  Even if you make an extra commit for each commit, you'd have to build off of the sha update commits in order for the sourcelink info to be correct.

**Comment by tmds** [comment]
> It is not ideal IMO that the sourcelink info would be off on all commits besides the release commit. Even if you make an extra commit for each commit, you'd have to build off of the sha update commits in order for the sourcelink info to be correct.

That's not what I meant.
I meant for the sourcelink info to only be present on the release commits.

**Comment by MichaelSimons** [comment]
> That's not what I meant.
> I meant for the sourcelink info to only be present on the release (that is: tagged) commits.

I see, so to build any non-tagged commits, you would have to specify the sourcelink parameters otherwise the info would exist and would be detected when no sourcelink parameters are specified.

**Comment by tmds** [comment]
Yes, the non-tagged commits wouldn't have a sourcelink info file.
If they have a `.git` folder, they can can still get the info from it.

**Comment by premun** [comment]
@tmds we have discussed this at length and finalized on following:
- We won't go the way of adding additional commits. We don't like several things about it:
   - Some commits will behave differently then others - some will have the metadata and some not, which might cause confusion. Some commits will still require additional info.
   - The VMR commit won't be self contained anymore as the second one will be referenced via source link.
   - The commit on the release tag will differ from the one we release through the Microsoft build.
   - It seems like we could easily make a mistake when branching or when adding hotfixes and MSRC fixes that would end up targeting a wrong commit if the information in the branch wouldn't get updated. This seems like a maintenance burden on the process.
- We have decided that the current approach of appending a metadata file on the release, or the alternative of having to have to provide the metadata is not a big burden because it only applies on non-git flows which are usually done by distro maintainers who already usually have the URL/SHA at hand and it seems it's not a big deal to require them to supply this.

Does this make sense or are there some points we missed where this new flow is problematic and not acceptable?
